app/mender.go

```diff
@@ -670,3 +686,195 @@ func verifyAndSetArtifactNameInProvides(
+		if bootstrapArtifactFileFound {
+			log.Debugf("Removing bootstrap Artifact %s", updatePath)
+			_ = os.Remove(updatePath)
+		}
+
+		// Initialize database on errors or file not found
+		if !databaseInitialized {
+			log.Debug("Bootstrap Artifact not installed, committing artifact_name 'unknown'")
+			if dbErr := m.Store.WriteTransaction(func(txn store.Transaction) error {
+				return datastore.CommitArtifactData(
+					txn,
+					"unknown",
+					"",
+					map[string]string{"artifact_name": "unknown"},
+					nil,
+				)
+			}); dbErr != nil {
+				if err != nil {
+					err = errors.Wrapf(err, dbErr.Error())
+				} else {
+					err = dbErr
+				}
+			}
+		}
+	}()
+
+	// Check if database already exists
+	_, sdErr := datastore.LoadStateData(s)
+	if sdErr == nil {
+		databaseInitialized = true
+	} else {
+		log.Debug("No state data stored")
+		p, pErr := datastore.LoadProvides(s)
+		if pErr == nil && len(p) > 0 {
+			databaseInitialized = true
+		} else {
+			log.Debug("No provides stored")
+			log.Debug("Database not initialized")
+		}
+	}
+
+	// Check if file exists
+	_, err = os.Stat(updatePath)
+	if err == nil {
+		log.Debugf("Bootstrap Artifact %s found", updatePath)
+		bootstrapArtifactFileFound = true
+	} else {
+		log.Debugf("Bootstrap Artifact %s not found", updatePath)
+	}
+
+	// Early exit, let the deferred function do the clean-ups
+	if databaseInitialized || !bootstrapArtifactFileFound {
+		if databaseInitialized && bootstrapArtifactFileFound {
+			log.Info(
+				"Bootstrap Artifact found but database already initialized, " +
+					"the Artifact will be ignored",
+			)
+		}
+		return nil
+	}
+
+	// Validate bootstrap Artifact
+	log.Debugf("Validating bootstrap Artifact from %s", updatePath)
+	dt, err := m.GetDeviceType()
+	if err != nil {
+		return errors.Wrap(err, "could not get device type")
+	}
+	artifactName, artifactGroup, provides, err := validateAndParseBootstrapArtifact(dt, updatePath)
+	if err != nil {
+		return errors.Wrap(err, "invalid bootstrap Artifact")
+	}
+
+	// Initialize database
+	log.Debugf("Initializing database from bootstrap Artifact")
+	err = m.Store.WriteTransaction(func(txn store.Transaction) error {
+		return datastore.CommitArtifactData(txn, artifactName, artifactGroup, provides, nil)
+	})
+	if err != nil {
+		return errors.Wrap(err, "Could not update database")
+	}
+
+	log.Info("Bootstrap Artifact installed successfully")
+	databaseInitialized = true
+
+	return nil
+
+}
+
+func validateAndParseBootstrapArtifact(
+	dt, path string,
+) (string, string, artifact.TypeInfoProvides, error) {
+	f, err := os.Open(path)
+	if err != nil {
+		return "", "", nil, err
+	}
+	defer f.Close()
+
+	ar := areader.NewReader(f)
+
+	ar.CompatibleDevicesCallback = func(devices []string) error {
+		for _, dev := range devices {
+			if dev == dt {
+				return nil
+			}
+		}
+		return fmt.Errorf(
+			"image (device types %v) not compatible with device %v",
```

Comment:
Q: Aren't they both strings?

Reply:
You are right, I'll modify. I took it from [the installer code](https://github.com/mendersoftware/mender/blob/1fd97067f6a50e7dc6446c63cc540b210a7ad3a2/installer/installer.go#L132), though.